src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs

```diff
@@ -90,72 +91,65 @@ internal static RuntimeGraph GetRuntimeGraph(IEnumerable targetFrameworks)
 
-
             tfi.RuntimeIdentifierGraphPath = GetPropertyValueOrNull(targetFrameworkInfo.Properties, ProjectBuildProperties.RuntimeIdentifierGraphPath);
 
-            if (targetFrameworkInfo.PackageReferences != null)
+            if (targetFrameworkInfo.Items is null)
             {
-                tfi.Dependencies.AddRange(
-                    targetFrameworkInfo.PackageReferences
-                        .Cast<IVsReferenceItem>()
-                        .Select(pr => ToPackageLibraryDependency(pr, cpvmEnabled)));
+                return tfi;
             }
 
-            if (targetFrameworkInfo is IVsTargetFrameworkInfo2 targetFrameworkInfo2)
+            if (targetFrameworkInfo.Items.TryGetValue(ProjectItems.PackageReference, out var packageReferences))
             {
-                if (targetFrameworkInfo2.PackageDownloads != null)
-                {
-                    tfi.DownloadDependencies.AddRange(
-                       targetFrameworkInfo2.PackageDownloads
-                           .Cast<IVsReferenceItem>()
-                           .SelectMany(ToPackageDownloadDependency));
-                }
+                tfi.Dependencies.AddRange(packageReferences.Select(pr => ToPackageLibraryDependency(pr, cpvmEnabled)));
+            }
 
-                if (cpvmEnabled && targetFrameworkInfo is IVsTargetFrameworkInfo3 targetFrameworkInfo3)
-                {
-                    if (targetFrameworkInfo3.CentralPackageVersions != null)
-                    {
-                        tfi.CentralPackageVersions.AddRange(
-                           targetFrameworkInfo3.CentralPackageVersions
-                               .Cast<IVsReferenceItem>()
-                               .Select(ToCentralPackageVersion)
-                               .Distinct(CentralPackageVersionNameComparer.Default)
-                               .ToDictionary(cpv => cpv.Name));
-                    }
-
-                    // Merge the central version information to the package information
-                    LibraryDependency.ApplyCentralVersionInformation(tfi.Dependencies, tfi.CentralPackageVersions);
-                }
+            if (targetFrameworkInfo.Items.TryGetValue("PackageDownload", out var packageDownloads))
+            {
+                tfi.DownloadDependencies.AddRange(packageDownloads.SelectMany(ToPackageDownloadDependency));
+            }
 
-                if (targetFrameworkInfo2.FrameworkReferences != null)
-                {
-                    PopulateFrameworkDependencies(tfi, targetFrameworkInfo2);
-                }
+            if (cpvmEnabled && targetFrameworkInfo.Items.TryGetValue("PackageVersion", out var centralPackageVersions))
+            {
+                tfi.CentralPackageVersions.AddRange(
+                    centralPackageVersions
+                       .Select(ToCentralPackageVersion)
+                       .Distinct(CentralPackageVersionNameComparer.Default)
```

Comment:
Do we still need the distinct here, given that we have a dictionary of id?

Reply:
The interface in this PR has `IReadOnlyDictionary<string, IReadOnlyList<IVsReferenceItem2>> Items` on the per-TFM data structure, but the key is the item type (PackageReference, PackageVersion, etc). The collection of items isn't a dictionary, just a list, like MSBuild fundamentally allowed. Now, we know that CPS and therefore dotnet/project-system collect items in per-type dictionaries. So maybe NuGet using a list, like MSBuild, is a leaky abstraction. I'm not sure how worried I should be about CPS/project-system in the future changing their mind and changing back to the MSBuild "shape" API, and treating them as lists that allow duplicate identifiers. This way NuGet won't need to change nomination API again.

Having said that, `Distinct` has a perf impact, so maybe we should avoid it anyway. We also de-duplicate in the `Collect*` targets, plus we know that for now CPS/project-system use a dictionary so they're not technically capable of providing us duplicates at this time.

Reply:
I think it's less risky to let it be at this point. 

Your call though. As far as I'm concerned, this is ok with no change.

Reply:
fwiw, there are other ways to improve the perf of the above.

For example, create a set with the right size, then add everything there instead of select and then distinct.